File: First Year/Second Semester/Graph Algorithms/Assignment2/Presentation/ConsoleUI.py
from Domain.UndirectedGraph import UndirectedGraph
from Domain.DirectedGraph import DirectedGraph, DirectedGraphException
from FileManager.FileManager import FileManager
from GraphGenerator.GraphRandomGenerator import GraphRandomGenerator


class ConsoleUI:
    def __init__(self):
        self.__directed_graph = DirectedGraph()
        self.__commands = {"1": self.__createRandomGraph, "2": self.__readGraphFromTextFile, "3": self.writeGraphToTextFile, "4": self.getVerticesNo,
                           "5": self.displayAllVertices, "6": self.testEdge, "7": self.displayInOutDegree, "8": self.displayOutboundEdges, "9": self.displayInboundEdges,
                           "10": self.getCost, "11": self.setCost, "12": self.addEdge, "13": self.removeEdge, "14": self.addVertex, "15": self.removeVertex,
                           "16": self.findLowestLengthPathBFS, "17": self.findStronglyConnectedComponents, "18": self.findBiconnectedComponents}

    def __createRandomGraph(self):
        vertices = input("Please input the number of vertices of the graph: ")
        edges = input("Please input the number of edges of the graph: ")

        try:
            self.__directed_graph = GraphRandomGenerator().generateGraph(int(vertices), int(edges))
        except ValueError as valueError:
            print(valueError)
        except DirectedGraphException as directedGraphException:
            print(directedGraphException)
        except Exception as exception:
            print(exception)

    def __readGraphFromTextFile(self):
        file_path = "Files/"
        file_path += input("Please input the name of the input file: ")
        try:
            self.__directed_graph = FileManager().read(file_path)
        except FileNotFoundError as fileNotFoundError:
            print(fileNotFoundError)
        except Exception as exception:
            print(exception)

    def writeGraphToTextFile(self):
        file_path = "Files/"
        file_path += input("Please input the name of the output file: ")
        try:
            FileManager().write(self.__directed_graph, file_path)
        except Exception as exception:
            print(exception)

    def getVerticesNo(self):
        vertices_no = self.__directed_graph.getVerticesNo()
        print("There are {} vertices in the graph.\n".format(str(vertices_no)))

    def displayAllVertices(self):
        print("The vertices of the graph are:")
        vertices_str = ""
        for vertex in self.__directed_graph.parseVertices():
            vertices_str += (str(vertex) + ", ")
        vertices_str = vertices_str[:-2]
        print(vertices_str)

    def testEdge(self):
        edge_x = input("Please input the source vertex: ")
        edge_y = input("Please input the destination vertex: ")
        try:
            message = "form" if self.__directed_graph.isEdge(int(edge_x), int(edge_y)) else "do not form"
            print("The two vertices " + message + " an edge.")
        except Exception as exception:
            print(exception)

    def displayInOutDegree(self):
        vertex = input("Please input a vertex: ")
        try:
            in_degree = self.__directed_graph.getInDegree(int(vertex))
            out_degree = self.__directed_graph.getOutDegree(int(vertex))
            print("The in degree of the given vertex is {}.".format(in_degree))
            print("The out degree of the given vertex is {}.".format(out_degree))
        except DirectedGraphException as directedGraphException:
            print(directedGraphException)
        except Exception as exception:
            print(exception)

    def displayOutboundEdges(self):
        try:
            vertex = int(input("Please input a vertex: "))
        except Exception as exception:
            print(exception)
            return
        print("The outbound edges of the vertex {} are:".format(vertex))
        try:
            for edge_y in self.__directed_graph.parseVerticesOut(vertex):
                print("({}, {})".format(vertex, edge_y))
        except DirectedGraphException as directedGraphException:
            print(directedGraphException)

    def displayInboundEdges(self):
        try:
            vertex = int(input("Please input a vertex: "))
        except Exception as exception:
            print(exception)
            return
        print("The inbound edges of the vertex {} is:".format(vertex))
        try:
            for edge_x in self.__directed_graph.parseVerticesIn(vertex):
                print("({}, {})".format(edge_x, vertex))
        except DirectedGraphException as directedGraphException:
            print(directedGraphException)

    def getCost(self):
        try:
            edge_x = int(input("Please input the source vertex of the edge: "))
            edge_y = int(input("Please input the destination vertex of the edge: "))
        except Exception as exception:
            print(exception)
            return
        try:
            cost = self.__directed_graph.getCost(edge_x, edge_y)
            print("The cost associated to the edge ({}, {}) is {}.".format(edge_x, edge_y, cost))
        except DirectedGraphException as directedGraphException:
            print(directedGraphException)

    def setCost(self):
        try:
            edge_x = int(input("Please input the source vertex of the edge: "))
            edge_y = int(input("Please input the destination vertex of the edge: "))
            cost = int(input("Please input the cost associated to the edge: "))
        except Exception as exception:
            print(exception)
            return
        try:
            self.__directed_graph.setCost(edge_x, edge_y, cost)
            print("The cost of the edge ({}, {}) was updated.".format(edge_x, edge_y))
        except DirectedGraphException as directedGraphException:
            print(directedGraphException)

    def addEdge(self):
        try:
            edge_x = int(input("Please input the source vertex of the edge: "))
            edge_y = int(input("Please input the destination vertex of the edge: "))
            cost = int(input("Please input the cost associated to the edge: "))
        except Exception as exception:
            print(exception)
            return
        try:
            self.__directed_graph.addEdge(edge_x, edge_y, cost)
            print("The edge ({}, {}) was added.".format(edge_x, edge_y))
        except DirectedGraphException as directedGraphException:
            print(directedGraphException)

    def removeEdge(self):
        try:
            edge_x = int(input("Please input the source vertex of the edge: "))
            edge_y = int(input("Please input the destination vertex of the edge: "))
        except Exception as exception:
            print(exception)
            return
        try:
            self.__directed_graph.removeEdge(edge_x, edge_y)
            print("The edge ({}, {}) was removed.".format(edge_x, edge_y))
        except DirectedGraphException as directedGraphException:
            print(directedGraphException)

    def addVertex(self):
        try:
            vertex = int(input("Please input a vertex: "))
        except Exception as exception:
            print(exception)
            return
        try:
            self.__directed_graph.addVertex(vertex)
            print("The vertex {} was added.".format(vertex))
        except DirectedGraphException as directedGraphException:
            print(directedGraphException)

    def removeVertex(self):
        try:
            vertex = int(input("Please input a vertex: "))
        except Exception as exception:
            print(exception)
            return
        try:
            self.__directed_graph.removeVertex(vertex)
            print("The vertex {} was removed.".format(vertex))
        except DirectedGraphException as directedGraphException:
            print(directedGraphException)

    def findLowestLengthPathBFS(self):
        try:
            source = int(input("Please input a source vertex: "))
            destination = int(input("Please input a destination vertex: "))
        except Exception as exception:
            print(exception)
            return
        try:
            distance, path = self.__directed_graph.breadthFirstSearch(source, destination)
            print("The length of the found path is: {}".format(distance))
            path_str = "The lowest length path is: "
            for node in path:
                path_str += "{} -> ".format(node)
            path_str = path_str[:-4]
            print(path_str)
        except DirectedGraphException as directedGraphException:
            print(directedGraphException)

    def findStronglyConnectedComponents(self):
        try:
            components = self.__directed_graph.findStronglyConnectedComponents()
            index = 1
            print("The graph has {} strongly connected components:".format(len(components)))

            for component in components:
                comp_str = "The strongly connected component no. {} is: ".format(index)
                for node in component:
                    comp_str += "{} -> ".format(node)
                comp_str = comp_str[:-4]
                print(comp_str)
                index += 1
        except DirectedGraphException as directedGraphException:
            print(directedGraphException)

    def findBiconnectedComponents(self):
        file_path = "Files/"
        print("An undirected graph must be loaded from an input file.")
        file_path += input("Please input the name of the input file: ")
        try:
            graph = FileManager().read(file_path, UndirectedGraph)
        except FileNotFoundError as fileNotFoundError:
            print(fileNotFoundError)
            return
        except Exception as exception:
            print(exception)
            return

        components = []

        components_no = graph.findBiconnectedComponents([], {}, [], components)
        if components_no > 1:
            print("\nNote that there are {} connected components in the graph".format(components_no))
        index = 1
        print("\nThe number of biconnected components is {}.".format(len(components)))
        for component in components:
            print("Biconnected component no. {} is: ".format(index))
            string = ""
            for node in component:
                string += "{} -> ".format(node)
            string = string[:-3]
            print(string)
            index += 1

    def __menu(self):
        print("\n" + "-" * 27 + " Graph Algorithms Assignment 1 Menu " + "-" * 27)
        print(" 1 - create a new random graph\n 2 - read a graph from a text file\n 3 - write the graph to a text file\n 4 - get the number of vertices in the graph")
        print(" 5 - display all the vertices in the graph\n 6 - test if there is an edge between two vertices\n 7 - display the in/out degree of a vertex")
        print(" 8 - display the outbound edges of a vertex\n 9 - display the inbound edges of a vertex\n10 - get cost of an edge\n11 - set cost of an edge")
        print("12 - add an edge\n13 - remove an edge\n14 - add a vertex\n15 - remove a vertex\n16 - given two vertices, finds a lowest length path between them, by using", end=" ")
        print("a forward breadth-first search from the starting vertex\n17 - find the strongly-connected components of the directed graph")
        print("18 - find the biconnected components of an undirected graph\n19 - exit the application")
        print("-" * 90)

    def run(self):
        running = True
        while running:
            try:
                self.__menu()
                command = input("command > ").strip()
                if command == "19":
                    running = False
                    break
                if command in self.__commands:
                    self.__commands[command]()
                else:
                    print("Invalid command.\n")
            except DirectedGraphException as directedGraphException:
                print("DirectedGraphException: " + str(directedGraphException))
            except Exception as exception:
                print("Unexpected exception occurred: " + str(exception))

File: First Year/Second Semester/Graph Algorithms/Assignment1CPP/ConsoleUI/ConsoleUI.cpp
//
// Created by Emanuel on 28.03.2021.
//

#include "ConsoleUI.h"
#include "../GraphGenerator/GraphRandomGenerator.h"
#include "../FileManager/FileManager.h"
#include "../TextTable/TextTable.h"

#include <iostream>

std::string ConsoleUI::getCommand() const {
    std::string command;
    std::cout << "command > ";
    std::getline(std::cin, command);
    std::cout << '\n';
    return command;
}

std::string ConsoleUI::getUserData(const std::string &message) const {
    std::string user_data;
    std::cout << message << "\n> ";
    std::getline(std::cin, user_data);
    std::cout << '\n';
    return user_data;
}

int ConsoleUI::mapCommand(const std::string &command) const {
    std::string commands[] = {"1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16"};
    int int_commands[] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15};

    for (size_t i = 0; i < 16; ++i)
        if (command == commands[i]) return int_commands[i];
    return -1;
}

void ConsoleUI::createRandomGraph() {
    std::string vertices = getUserData("Please input the number of vertices of the graph: ");
    std::string edges = getUserData("Please input the number of edges of the graph: ");

    if (!isPositiveInteger(vertices) || !isPositiveInteger(edges)) {
        std::cout << "Invalid vertices/edges number.\n";
        return;
    }

    try {
        GraphRandomGenerator generator;
        directed_graph = generator.generateGraph(stoi(vertices), stoi(edges));
    }
    catch (std::string error) {
        std::cout << error << '\n';
    }
}

void ConsoleUI::readGraphFromTextFile() {
    std::string file_path = "../Files/";
    file_path += getUserData("Please input the name of the input file: ");

    try {
        FileManager file_manager;
        directed_graph = file_manager.read(file_path);
    }
    catch (std::string error) {
        std::cout << error << '\n';
    }
}

void ConsoleUI::writeGraphToTextFile() {
    std::string file_path = "../Files/";
    file_path += getUserData("Please input the name of the output file: ");

    try {
        FileManager file_manager;
        file_manager.write(directed_graph, file_path);
    }
    catch (std::string error) {
        std::cout << error << '\n';
    }
}

void ConsoleUI::getVerticesNo() const {
    int vertices_no = directed_graph.getVerticesNo();
    std::cout << "There are " << vertices_no << " vertices in the graph.\n";
}

void ConsoleUI::displayAllVertices() const {
    std::cout << "The vertices of the graph are:\n";
    std::string vertices_str = "";

    auto it = directed_graph.parseVertices();
    while (it.isValid()) {
        vertices_str += std::to_string(it.getCurrent()) + ", ";
        it.next();
    }
    vertices_str.pop_back();
    vertices_str.pop_back();
    std::cout << vertices_str << '\n';
}

void ConsoleUI::testEdge() {
    std::string edge_x = getUserData("Please input the source vertex: ");
    std::string edge_y = getUserData("Please input the destination vertex: ");

    if (!isPositiveInteger(edge_x) || !isPositiveInteger(edge_y)) {
        std::cout << "Invalid vertices.\n";
        return;
    }

    try {
        std::string message = directed_graph.isEdge(stoi(edge_x), stoi(edge_y)) ? "form" : "do not form";
        std::cout << "The two vertices " << message << " an edge.\n";
    }
    catch (std::string error) {
        std::cout << error << '\n';
    }
}

void ConsoleUI::displayInOutDegree() {
    std::string vertex = getUserData("Please input a vertex: ");
    if (!isPositiveInteger(vertex)) {
        std::cout << "Invalid vertex.\n";
        return;
    }

    try {
        int in_degree = directed_graph.getInDegree(stoi(vertex));
        int out_degree = directed_graph.getOutDegree(stoi(vertex));
        std::cout << "The in degree of the given vertex is " << in_degree << ".\n";
        std::cout << "The out degree of the given vertex is " << out_degree << ".\n";
    }
    catch (std::string error) {
        std::cout << error << '\n';
    }
}

void ConsoleUI::displayOutboundEdges() {
    std::string vertex = getUserData("Please input a vertex: ");
    if (!isPositiveInteger(vertex)) {
        std::cout << "Invalid vertex.\n";
        return;
    }

    std::cout << "The outbound edges of the vertex " << vertex << " are:\n";
    try {
        auto it = directed_graph.parseVerticesOut(stoi(vertex));
        while (it.isValid()) {
            std::cout << "(" << vertex << ", " << std::to_string(it.getCurrent()) << ")\n";
            it.next();
        }
    }
    catch (std::string error) {
        std::cout << error << '\n';
    }
}

void ConsoleUI::displayInboundEdges() {
    std::string vertex = getUserData("Please input a vertex: ");
    if (!isPositiveInteger(vertex)) {
        std::cout << "Invalid vertex.\n";
        return;
    }

    std::cout << "The inbound edges of the vertex " << vertex << " are:\n";
    try {
        auto it = directed_graph.parseVerticesIn(stoi(vertex));
        while (it.isValid()) {
            std::cout << "(" << std::to_string(it.getCurrent()) << ", " << vertex << ")\n";
            it.next();
        }
    }
    catch (std::string error) {
        std::cout << error << '\n';
    }
}

void ConsoleUI::getCost() {
    std::string edge_x = getUserData("Please input the source vertex of the edge: ");
    std::string edge_y = getUserData("Please input the destination vertex of the edge: ");

    if (!isPositiveInteger(edge_x) || !isPositiveInteger(edge_y)) {
        std::cout << "Invalid vertices.\n";
        return;
    }

    try {
        int cost = directed_graph.getCost(stoi(edge_x), stoi(edge_y));
        std::cout << "The cost associated to the edge (" << edge_x << ", " << edge_y << ") is " << cost << ".\n";
    }
    catch (std::string error) {
        std::cout << error << '\n';
    }
}

void ConsoleUI::setCost() {
    std::string edge_x = getUserData("Please input the source vertex of the edge: ");
    std::string edge_y = getUserData("Please input the destination vertex of the edge: ");
    std::string cost = getUserData("Please input the cost associated to the edge: ");

    if (!isPositiveInteger(edge_x) || !isPositiveInteger(edge_y)) {
        std::cout << "Invalid vertices.\n";
        return;
    }
    if (!isPositiveInteger(cost)) {
        std::cout << "Invalid edge cost.\n";
        return;
    }

    try {
        directed_graph.setCost(stoi(edge_x), stoi(edge_y), stoi(cost));
        std::cout << "The cost associated to the edge (" << edge_x << ", " << edge_y << ") 1was update.\n";
    }
    catch (std::string error) {
        std::cout << error << '\n';
    }
}

void ConsoleUI::addEdge() {
    std::string edge_x = getUserData("Please input the source vertex of the edge: ");
    std::string edge_y = getUserData("Please input the destination vertex of the edge: ");
    std::string cost = getUserData("Please input the cost associated to the edge: ");

    if (!isPositiveInteger(edge_x) || !isPositiveInteger(edge_y)) {
        std::cout << "Invalid vertices.\n";
        return;
    }
    if (!isPositiveInteger(cost)) {
        std::cout << "Invalid edge cost.\n";
        return;
    }

    try {
        directed_graph.addEdge(stoi(edge_x), stoi(edge_y), stoi(cost));
        std::cout << "The edge (" << edge_x << ", " << edge_y << ") was added.\n";
    }
    catch (std::string error) {
        std::cout << error << '\n';
    }
}

void ConsoleUI::removeEdge() {
    std::string edge_x = getUserData("Please input the source vertex of the edge: ");
    std::string edge_y = getUserData("Please input the destination vertex of the edge: ");

    if (!isPositiveInteger(edge_x) || !isPositiveInteger(edge_y)) {
        std::cout << "Invalid vertices.\n";
        return;
    }

    try {
        directed_graph.removeEdge(stoi(edge_x), stoi(edge_y));
        std::cout << "The edge (" << edge_x << ", " << edge_y << ") was removed.\n";
    }
    catch (std::string error) {
        std::cout << error << '\n';
    }
}

void ConsoleUI::addVertex() {
    std::string vertex = getUserData("Please input a vertex: ");
    if (!isPositiveInteger(vertex)) {
        std::cout << "Invalid vertex.\n";
        return;
    }

    try {
        directed_graph.addVertex(stoi(vertex));
        std::cout << "The vertex " << vertex << " was added.\n";
    }
    catch (std::string error) {
        std::cout << error << '\n';
    }
}

void ConsoleUI::removeVertex() {
    std::string vertex = getUserData("Please input a vertex: ");
    if (!isPositiveInteger(vertex)) {
        std::cout << "Invalid vertex.\n";
        return;
    }

    try {
        directed_graph.removeVertex(stoi(vertex));
        std::cout << "The vertex " << vertex << " was removed.\n";
    }
    catch (std::string error) {
        std::cout << error << '\n';
    }
}

void ConsoleUI::menu() const {
    TextTable table {2};
    table.addHeader(2, "COMMAND", "COMMAND DESCRIPTION");
    std::string commands[] = {"1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16"};
    std::string command_descriptions[] = {"create a new random graph", "read a graph from a text file", "write the graph to a text file",
                                          "get the number of vertices in the graph", "display all the vertices in the graph",
                                          "test if there is an edge between two vertices", "display the in/out degree of a vertex",
                                          "display the outbound edges of a vertex", "display the inbound edges of a vertex", "get cost of an edge",
                                          "set cost of an edge", "add an edge", "remove an edge", "add a vertex", "remove a vertex",
                                          "exit the application"};
    for (size_t i = 0; i < 16; ++i)
        table.addRow(2, commands[i].c_str(), command_descriptions[i].c_str());
    std::cout << "Application Menu:\n" << table << '\n';
}

void ConsoleUI::run() {
    bool running = true;
    while (running) {
        try {
            menu();

            int command = mapCommand(getCommand());
            switch (command) {
                case 0:
                    createRandomGraph();
                    break;
                case 1:
                    readGraphFromTextFile();
                    break;
                case 2:
                    writeGraphToTextFile();
                    break;
                case 3:
                    getVerticesNo();
                    break;
                case 4:
                    displayAllVertices();
                    break;
                case 5:
                    testEdge();
                    break;
                case 6:
                    displayInOutDegree();
                    break;
                case 7:
                    displayOutboundEdges();
                    break;
                case 8:
                    displayInboundEdges();
                    break;
                case 9:
                    getCost();
                    break;
                case 10:
                    setCost();
                    break;
                case 11:
                    addEdge();
                    break;
                case 12:
                    removeEdge();
                    break;
                case 13:
                    addVertex();
                    break;
                case 14:
                    removeVertex();
                    break;
                case 15:
                    running = false;
                    break;
                default:
                    std::cout << "Invalid command.\n";
                    break;
            }
        }
        catch (...) {
            std::cout << "Unexpected error occurred.\n";
        }
    }
}

bool ConsoleUI::isPositiveInteger(const std::string& string) const {
    for (size_t i = 0; i < string.length(); ++i)
        if (!isdigit(string[i])) return false;
    return true;
}
